src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,499 @@
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Parameters for the combined Extract-Only, Expand-Only, or Extract-then-Expand
+ * operations of the HMAC-based Key Derivation Function (HKDF). The HKDF
+ * function is defined in <a href="http://tools.ietf.org/html/rfc5869">RFC
+ * 5869</a>.
+ * <p>
+ * In the Extract-Only and Extract-then-Expand cases, the {@code addIKM} and
+ * {@code addSalt} methods may be called repeatedly (and chained). This provides
+ * for use-cases where a {@code SecretKey} may reside on an HSM and not be
+ * exportable. The caller may wish to provide a label (or other components) of
+ * the IKM without having access to the portion stored on the HSM. The same
+ * feature is available for salts.
+ * <p>
+ * The above feature is particularly useful for "labeled" HKDF Extract used in
+ * TLS 1.3 and HPKE, where the IKM consists of concatenated components, which
+ * may include both byte arrays and (possibly non-extractable) secret keys.
+ * <p>
+ * Examples:
+ * {@snippet lang = java:
+ *
+ * KDFParameterSpec kdfParameterSpec =
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(ikmPart1)
+ *                              .addIKM(ikmPart2)
+ *                              .addSalt(salt).extractOnly();
+ *
+ *
+ *}
+ * {@snippet lang = java:
+ *
+ * KDFParameterSpec kdfParameterSpec = HKDFParameterSpec.expandOnly(prk,
+ * info, 32);
+ *
+ *}
+ * {@snippet lang = java:
+ *
+ * KDFParameterSpec kdfParameterSpec =
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(ikm)
+ *                              .addSalt(salt).thenExpand(info, 42);
+ *
+ *}
+ *
+ * @since 23
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public interface HKDFParameterSpec extends KDFParameterSpec {
+
+    /**
+     * This {@code Builder} builds {@code Extract} and {@code ExtractThenExpand}
+     * objects.
+     * <p>
+     * The {@code Builder} is initialized via the {@code ofExtract} method of
+     * {@code HKDFParameterSpec}. As stated in the class description,
+     * {@code addIKM} and/or {@code addSalt} may be called as needed. Finally,
+     * the object is "built" by calling either {@code extractOnly} or
+     * {@code thenExpand} for {@code Extract} and {@code ExtractThenExpand}
+     * use-cases respectively.
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+    final class Builder {
+
+        List<SecretKey> ikms = new ArrayList<>();
+        List<SecretKey> salts = new ArrayList<>();
+
+        Builder() {}
+
+        /**
+         * Creates a {@code Builder} for an {@code Extract}.
+         *
+         * @return a {@code Builder} to mutate
+         */
+        private Builder createBuilder() {
+            return this;
+        }
+
+        /**
+         * Builds an {@code Extract}.
+         *
+         * @return an immutable {@code Extract}
+         */
+        public Extract extractOnly() {
```

Comment:
I think you can expand the method description to say that it builds the Extract method from the current state of the builder.